### DISCUSSION TITLE
Can firejail relay SIGHUP and SIGUSR2 to the underlying process?

### DISCUSSION THREAD

**Discussion by crocket** [netblue30/firejail#4551] [Q&A] [answered]
I sandboxed an OpenRC init script service with firejail. I noticed that I need to rotate log by sending SIGUSR2 to the process.

However, firejail catches SIGUSR2 instead.

**Comment by smitsohu** [comment]
Just describing how it works right now....

If you send SIGUSR2 to the Firejail main process, it kills everything immediately, without processes inside the sandbox having a chance to handle the signal.

If you send SIGUSR2 to the child of the main process, let's call it the sandbox process, nothing will happen. That's because the sandbox process is pid 1 of a pid namespace, and the kernel silently drops all signals for which pid 1 does not have a signal handler (there are few exceptions). At the moment the sandbox process has signal handlers only for SIGINT and SIGTERM.

**Comment by smitsohu** [comment]
That said, broadcasting these signals in the sandbox is of course possible, we do something similar already with SIGINT and SIGTERM.

**Reply by crocket** [reply]
Can an init script join a sandbox, find the PID of the underlying process, and send the signal?

**Reply by rusty-snake** [reply]
> Can an init script join a sandbox, find the PID of the underlying process, and send the signal?

If `seccomp` allows the necessary syscalls and `private-bin` the necessary program for shell scripts, yes. But why do you want to join the sandbox for that? Does it not work to parse `firejail --tree` (or pstree or anything like that) and send it from outside to the process?

**Reply by smitsohu** [reply]
Or you just do

`firejail --join=name kill -SIGUSR2 -1`

The pid namespace will limit the impact to the sandbox.

**Reply by crocket** [reply]
`firejail --tree` is too complex. I can't parse it.

**Reply by crocket** [reply]
```
$ pstree -pTA firejail-PID
firejail(4804)---firejail(4813)---dnscrypt-proxy(4822)
$ pstree -pTA firejail-PID
firejail(7789)---firejail(7791)---librewolf(7802)-+-Web Content(8218)
                                                  |-Web Content(8410)
                                                  |-Web Content(9679)
                                                  `-WebExtensions(8131)
```
I can parse this with regular expressions or parsing expression grammars(PEG). I prefer PEG because it's readable by humans.

I can probably write a program called `firejail-kill`.

```
firejail-kill firejail-PID SIGNAL
```

**Reply by rusty-snake** [reply]
> `firejail --tree` is too complex. I can't parse it.

If you know the command line ...
```bash
CMD="/usr/lib64/firefox/firefox --new-window"
firejail --tree | awk -F : "/^    [0-9]+:${USER:-$(id -un)}::${CMD//\//\/}/ { print \$1 } " | xargs kill -SIGUSR1 --
```
(can certainly be improved)

**Reply by crocket** [reply]
Parsing `pstree -pTA firejail-PID` is simpler.

**Comment by crocket** [comment, marked as answer]
I think

```
firejail --join=name kill -SIGUSR2 -1
```
or parsing `pstree -pAT firejail-main-PID` is going to do the job.
